File: 2021/07/crabs.ts
import * as fs from 'fs'
import * as _ from 'lodash'

const input = fs.readFileSync('input/07.txt').toString().split(',').map(x => parseInt(x))

function fuel(x: number, x1: number): number {
    return Math.abs(x - x1)
}

function fuel2(x: number, x1: number): number {
    const distance = Math.abs(x - x1)
    const [max, rem] = distance % 2 == 0 ? [distance, 0] : [distance - 1, distance]

    return (1 + max) * Math.floor(distance / 2) + rem

}

function minimum(input: number[], fuel: (x: number, x1: number) => number): number {
    const sorted = input.sort()
    const min = sorted[0]
    const max = sorted.slice(-1)[0]

    return _.range(min, max).map(position => input.map(x => fuel(position, x)).reduce((m, x) => m + x)).sort()[0]
}


console.log(minimum(input, fuel))

console.log(minimum(input, fuel2))

console.log(fuel2(5,16))
console.log(fuel2(2,5))
console.log(fuel2(1,5))

File: 2021/07/crabs.rb
def fuel(x, x1) = (x - x1).abs

def fuel2(x, x1)
  distance = (x - x1).abs
  max, rem = distance.even? ? [distance, 0] : [distance - 1, distance]

  (1 + max) * (distance / 2) + rem
end

input = (STDIN.tty? ? DATA : STDIN).read.split(',').map(&:to_i)

min, *_, max = input.sort

p (min..max).map { |target| input.map { fuel(_1, target) }.sum }.min
p (min..max).map { |target| input.map { fuel2(_1, target) }.sum }.min

__END__
16,1,2,0,4,2,7,1,2,14
